### ISSUE TITLE
APP packet payload length?

### ISSUE THREAD

**Issue by wowaser** [ultravideo/uvgRTP#126]
Hello, I've been getting some mysterious crashes, possibly related to the APP packets. Specifically when I am trying to get the payload from it to a string. 

The payload is *uint8_t**, and the contents of the header are the following:
![image](https://user-images.githubusercontent.com/70014877/172241163-5ad3fd60-ed94-4327-b336-243d0d24dc28.png)
I was wondering if the *length* is the header, packet, or the payload length? 
What is the best way of extracting the payload to a string?
So far I've been trying 
```cpp
void hook(std::shared_ptr<uvgrtp::frame::rtcp_app_packet> packet)
{
    std::string str = (char*)packet->payload;
    //OR
    std::string str(packet->payload, packet->payload+packet->header.length);
    //OR
    std::string str(reinterpret_cast<char*>(packet->payload));
    //etc
}
```

Thanks

**Comment by jrsnen** [comment]
Hello.

The meaning of length value can be found in RFC 3550 which uvgRTP tries to follow (I don't remember off the top of my head). 

As for what is the best way, I have to say I'm not sure since I have not used APP packets before. I'd try with memcpy to a same type pointer that was saved to payload.

**Comment by jrsnen** [comment]
Related: #35

**Comment by jrsnen** [comment]
Now that #35 has been fixed, the length field in RTCP header is measured in 32-bit words -1. I added a handy helper function to rtcp.hh called rtcp_length_in_bytes which you can use to get the amount of bytes in the whole rtcp message. Payload length is this length - the size of header(4) - size of ssrc(4) - size of APP name (4) so its basically size of the packet minus 12 (or -3 if measured in 32-bit words). There may be issues if the payload length is not exactly measured in 32-bit words as discussed in #139

To recap, if APP payloads are not measured in 32-bit words, you could for example use 2 first bits to indicate how many bytes of padding you have (0-3).

I will add a check that prevents sending incorrect length messages. Does this answer your question?

**Comment by jrsnen** [comment]
It also would be super nice if someone could document this somewhere. I'm too busy with other stuff, so I'm not adding the doc-tag, as I would never get to it.

**Comment by jrsnen** [comment]
I added the tag anyway if someone wants to do this.

**Comment by jrsnen** [comment]
![image](https://user-images.githubusercontent.com/6498638/178420854-1c939f08-7191-48eb-b415-8570d24a2c4d.png)

**Comment by jrsnen** [comment]
I was actually thinking that the payload length could be added as a field to app packet struct in frame.hh

**Comment by jrsnen** [comment]
Added an APP payload length field to the packet in f55c4578a835f7c4ff39e05a400c02349a22e5ab, so users don't have to calculate it themselves.

Field can be found here: https://github.com/ultravideo/uvgRTP/blob/f55c4578a835f7c4ff39e05a400c02349a22e5ab/include/uvgrtp/frame.hh#L138
